store/src/main/java/com/dropbox/android/external/store4/StoreResponse.kt

```diff
@@ -96,12 +110,27 @@ enum class ResponseOrigin {
      * [StoreResponse] is sent from the cache
      */
     Cache,
+
     /**
      * [StoreResponse] is sent from the persister
      */
     Persister,
+
     /**
      * [StoreResponse] is sent from a fetcher,
      */
     Fetcher
 }
+
+sealed class FetcherResult<T : Any> {
+    data class Data<T : Any>(val value: T) : FetcherResult<T>()
+    sealed class Error<T : Any> : FetcherResult<T>() {
+        data class Exception<T : Any>(val error: Throwable) : Error<T>()
+        data class Message<T : Any>(val message: String) : Error<T>()
+    }
+}
+
+fun <T> StoreResponse.Error<T>.doThrow(): Nothing = when (this) {
+    is StoreResponse.Error.Exception -> throw error
+    is StoreResponse.Error.Message -> throw IllegalStateException(message)
```

Comment:
RuntimeException is probably more appropriate here since we don't really know if it is illegal state

Reply:
alternatively, we can have an error class coming from Store.

Reply:
Going with runtime for now